File: WINGS_AI_Bootcamp/Haskell/Odometer.hs
import Data.Char (digitToInt)

isValid :: Int -> Bool
isValid n
  | n < 10 = True
  | otherwise =
      let digits = map digitToInt (show n)
          pairs = zip digits (tail digits)
       in all (\(a,b) -> a < b) pairs

getLimits :: Int -> (Int, Int)
getLimits n =
  let limit = "123456789"
      size = length (show n)
      start = read (take size limit) :: Int
      end = read (drop (length limit - size) limit) :: Int
   in (start, end)

nextReading :: Int -> Int
nextReading reading =
  let (start, limit) = getLimits reading
      next n
        | n > limit = start
        | isValid n = n
        | otherwise = next (n + 1)
   in if reading == limit then start else next (reading + 1)

nextReadingK :: Int -> Int -> Int
nextReadingK reading 0 = reading
nextReadingK reading k = nextReadingK (nextReading reading) (k - 1)

prevReading :: Int -> Int
prevReading reading =
  let (start, limit) = getLimits reading
      prev n
        | n < start = limit
        | isValid n = n
        | otherwise = prev (n - 1)
   in if reading == start then limit else prev (reading - 1)

prevReadingK :: Int -> Int -> Int
prevReadingK reading 0 = reading
prevReadingK reading k = prevReadingK (prevReading reading) (k - 1)

distance :: Int -> Int -> Int
distance a b
  | length (show a) /= length (show b) = -1
  | otherwise = go a 0
  where
    (_, limit) = getLimits a
    go curr count
      | count > 10^(length (show a)) = -1
      | curr == b = count
      | otherwise = go (nextReading curr) (count + 1)

-- Examples
main :: IO ()
main = do
  print $ nextReading 2468
  print $ nextReadingK 2468 3
  print $ prevReading 2468
  print $ prevReadingK 2468 3
  print $ distance 2468 2469

File: WINGS_AI_Bootcamp/scala_Ymeg/odometer.scala
def isValid(n: Int): Boolean =
  if n < 10 then true
  else if (n / 10) % 10 < n % 10 then isValid(n / 10)
  else false

def getLimits(n: Int): (Int, Int) =
  val LIMIT = "123456789"
  val size = n.toString.length
  (LIMIT.take(size).toInt, LIMIT.takeRight(size).toInt)

def nextReading(reading: Int): Int =
  val (start, limit) = getLimits(reading)
  if reading == limit then start
  else
    var r = reading + 1
    while !isValid(r) do r += 1
    r

def nextReadingK(reading: Int, step: Int = 1): Int =
  var r = reading
  for _ <- 1 to step do
    r = nextReading(r)
  r

def prevReading(reading: Int, step: Int = 1): Int =
  val (start, limit) = getLimits(reading)
  if reading == start then limit
  else
    var r = reading - 1
    while !isValid(r) do r -= 1
    r

def distance(a: Int, b: Int): Int =
  if a.toString.length != b.toString.length then -1
  else
    var count = 0
    var r = a
    while r != b do
      r = nextReading(r)
      count += 1
    count
